Frameworks/CoreGraphics/CGGradient.mm

```diff
@@ -18,186 +18,170 @@
-        UIColor* curColor = [components objectAtIndex:i];
+        if (locations) {
+            std::copy(locations, locations + count, std::back_inserter(_stopLocations));
+            return;
+        }
 
-        if (curColor) {
-            const __CGColorQuad* color = [curColor _getColors];
-            memcpy(&_components[i * componentCount], color, sizeof(CGFloat) * componentCount);
-        } else {
-            memset(&_components[i * componentCount], 0, sizeof(CGFloat) * componentCount);
+        for (size_t i = 0; i < count; i++) {
+            _stopLocations[i] = (CGFloat)i / (CGFloat)(count - 1);
         }
     }
+};
 
-    _locations = new float[count];
+void __CGGradient::InitWithColorComponents(const CGFloat* components, const CGFloat* locations, size_t count) {
+    size_t componentCount = CGColorSpaceGetNumberOfComponents(_colorSpace) + 1;
+    _colorComponents.reserve(count * componentCount);
+    std::copy(components, components + (count * componentCount), std::back_inserter(_colorComponents));
+    SetLocations(locations, count);
+}
 
-    if (locations) {
-        memcpy(_locations, locations, sizeof(float) * count);
-    } else {
-        for (int i = 0; i < count; i++) {
-            _locations[i] = (float)i / (float)(count - 1);
-        }
+void __CGGradient::InitWithColors(CFArrayRef components, const CGFloat* locations) {
+    size_t componentCount = CGColorSpaceGetNumberOfComponents(_colorSpace) + 1;
+    CFIndex count = CFArrayGetCount(components);
+    _colorComponents.reserve(count * componentCount);
+    CGFloat defaultColor[4] = { 0, 0, 0, 0 };
+
+    const CGFloat* color;
+    for (CFIndex i = 0; i < count; i++) {
+        CGColorRef curColor = (CGColorRef)CFArrayGetValueAtIndex(components, i);
+        const CGFloat* color = curColor ? CGColorGetComponents(curColor) : defaultColor;
+        _colorComponents.insert(_colorComponents.begin() + i * componentCount, color, color + componentCount);
     }
 
-    _count = count;
+    SetLocations(locations, count);
 }
 
 /**
  @Status Interoperable
 */
-CGGradientRef CGGradientCreateWithColorComponents(CGColorSpaceRef colorSpace,
-                                                  const float* components,
-                                                  const float* locations,
-                                                  size_t count) {
-    CGGradientRef ret = new __CGGradient();
-    ret->initWithColorComponents(components, locations, count, colorSpace);
-
-    return ret;
+CFTypeID CGGradientGetTypeID() {
+    return __CGGradient::GetTypeID();
 }
 
 /**
  @Status Interoperable
 */
-CGGradientRef CGGradientCreateWithColors(CGColorSpaceRef colorSpace, CFArrayRef colors, const float* locations) {
-    CGGradientRef ret = new __CGGradient();
-    ret->initWithColors(colors, locations, colorSpace);
-
-    return ret;
+void CGGradientRelease(CGGradientRef gradient) {
+    RETURN_IF(!gradient);
+    CFRelease(static_cast<CFTypeRef>(gradient));
```

Comment:
it's needed.

---
In reply to: [106461735](https://github.com/Microsoft/WinObjC/pull/2249#discussion_r106461735) [](ancestors = 106461735)